File: src/main/scala/fungsi/src/Token.scala
package fungsi.src

sealed abstract class TokenType

class Token(val token: TokenType, val content: String):
  override def toString: String = "{ " + token + ", " + content + " }"
  
  def toList: List[Token] = List(this)

  def equals(that: Token): Boolean = this.content == that.content

object Token:
  trait Term
  trait Nonterm

  case object DEF extends TokenType with Term // def
  case object BIND extends TokenType with Nonterm // :=
  case object SEMICOLON extends TokenType with Nonterm // ;
  case object COMMA extends TokenType with Nonterm // ,
  case object NAME extends TokenType with Term
  case object ATOM extends TokenType with Term // :{name}
  case object LAMBDA extends TokenType with Term// @
  case object DELIMITER extends TokenType with Nonterm // |
  case object QUOTE extends TokenType with Nonterm // $
  case object RETURN extends TokenType with Nonterm // ^
  case object NUM extends TokenType with Term
  case object CHAR extends TokenType with Term
  case object STRING extends TokenType with Term
  case object LPAREN extends TokenType with Term// ( or [
  case object RPAREN extends TokenType with Nonterm // ) or ]
  case object EOF extends TokenType with Nonterm

File: src/token.rs
use std::collections::HashMap;
use std::fmt;
use std::fmt::Formatter;

#[derive(fmt::Debug, Clone, Copy)]
pub enum TokenType {
    Nil,       // nil
    Object,    // object
    Bind,      // :=
    Arrow,     // <-
    Dot,       // .
    Comma,     // ,
    Name,      // xyz you know
    String,    // ""
    Atom,      // :atom
    Lambda,    // @
    Delimiter, // |
    Return,    // ^
    Num,       // 123 you know
    LParen,    // ( or [
    RParen,    // ) or ]
    Eof,
    // Comment    NB
}

impl TokenType {
    pub fn is_gen_sym(self) -> bool {
        use TokenType::*;
        matches!(
            self,
            Nil | Object | Name | Atom | Lambda | Num | String | LParen
        )
    }
}

pub fn make_keywords() -> HashMap<&'static str, TokenType> {
    let mut keywords = HashMap::new();
    keywords.insert("nil", TokenType::Nil);
    keywords.insert("object", TokenType::Object);
    keywords.insert("<-", TokenType::Arrow);
    keywords
}

impl fmt::Display for TokenType {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}

#[derive(Debug, Clone)]
pub struct Token {
    pub(crate) token: TokenType,
    pub(crate) content: String,
    pub(crate) span: (i32, i32),
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        write!(f, "{{ {} , {} }}", self.token, self.content)
    }
}

impl PartialEq for Token {
    fn eq(&self, other: &Self) -> bool {
        self.content == other.content
    }
}
